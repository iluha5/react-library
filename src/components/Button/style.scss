@import "../../variables";

$btn-base-color: #3388e7;

.Btn {
    border-radius: 3px;
    border: 2px solid transparent;
    background: none;
    font-size: 14px;
    font-weight: 300;
    transition: 0.2s all;
    outline: none;

    &_size {
        &_m {
            padding: 7px    30px;
        }

        &_xs {
            padding: 3px 27px 4px 27px;
        }
    }

    &_type {
        &_primary {
            background: $primary-color;
            color: $text-base-color;
            &:hover {
                background-color: transparent;
                color: $primary-color;
                border-color: $primary-color;
            }
        }

        &_secondary {
            background-color: $secondary-color;
            color: $text-base-color;
            &:hover {
                background-color: transparent;
                color: $primary-color;
                border-color: $primary-color;
            }
        }

        &_success {
            background-color: $success-color;
            color: $text-base-color;
            &:hover {
                background-color: transparent;
                color: $primary-color;
                border-color: $primary-color;
            }
        }

        &_link {
            color: $link-color;
            &:hover {
                color: #00FFFB;
            }
        }

        &_fake {
            background: $fake-color;
            color: $text-base-color;
            &:hover {
                background-color: transparent;
                color: $secondary-color;
                border-color: $secondary-color;
            }
        }
    }

    &:hover,
    &:focus,
    &:active {
        outline: none;
        cursor: pointer;
    }

    &:disabled,
    &:disabled:focus,
    &:disabled:hover,
    &:disabled:active {
        background-color: $disabled-color;
        border-color: $disabled-color;
        color: rgba(255, 255, 255, 0.3);
        cursor: not-allowed;
    }
}

.Btn-Label {
    display: inline-block;
}




