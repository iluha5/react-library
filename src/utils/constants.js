export const NOTIFICATION_ERROR = 'NOTIFICATION_ERROR';
export const NOTIFICATION_PASSED = 'NOTIFICATION_PASSED';

// Regex
// export const REGEX_PASSWORD = /^(?=.*[A-Za-z])(?=.*\d)(?=.*[@$!%*#?&])[A-Za-z\d@$!%*#?&]{8,32}$/; // eslint-disable-line
export const REGEX_PASSWORD = /^((?=.*[0-9])|(?=.*[@$!%*#?&,.\-+()\[\]{}_\/]))(?=.*[a-z])(?=.*[A-Z])([a-zA-Z0-9@$!%*#?&,.\-+()\[\]{}_\/]){8,32}$/; // eslint-disable-line
export const REGEX_EMAIL = /^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|((([a-zA-Z0-9]+[a-zA-Z\-0-9]+[a-zA-Z0-9]\.)|([a-zA-Z0-9]{1,}\.))+[a-zA-Z]{2,10}))$/; // eslint-disable-line
export const REGEX_NICKNAME = /^([a-zA-Z])[a-zA-Z0-9_-]{0,31}$/; // eslint-disable-line
export const REGEX_NAME = /^([a-zA-Z])[a-zA-Z0-9]{0,63}$/; // eslint-disable-line
