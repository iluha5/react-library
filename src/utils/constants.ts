export const NOTIFICATION_ERROR = 'NOTIFICATION_ERROR';
export const NOTIFICATION_PASSED = 'NOTIFICATION_PASSED';

export enum notificationTypes {
    ACHIEVEMENT = 'ACHIEVEMENT',
    PASSED = 'PASSED',
    LAB_STOP_ERROR = 'LAB_STOP_ERROR',
    LAB_START_ERROR = 'LAB_START_ERROR',
    NOTIFICATION_ERROR = 'NOTIFICATION_ERROR',
}

//
// export enum notificationClassSuffixes {
//     ERROR = 'Error',
//         PASSING = 'Passing',
//         ACHIEVEMENT = 'Achievement',
//         DEFAULT = ''
// }


export const TIME_OVER = 'TIME_OVER';
export const TEN_SECONDS_LEFT = 'TEN_SECONDS_LEFT';

// auth
export const PASSWORD_INPUT_MAX_LENGTH = 32;
export const NAME_INPUT_MAX_LENGTH = 64;
export const EMAIL_NOT_VERIFIED_MESSAGE = 'Email is not verified';

// HTTP codes
export enum httpResponseCodes {
    EMPTY = 0,
    HTTP_401 = 401,
    HTTP_422 = 422,
}

// Regex
export const REGEX_PASSWORD = /^((?=.*[0-9])|(?=.*[@$!%*#?&,.\-+()\[\]{}_\/]))(?=.*[a-z])(?=.*[A-Z])([a-zA-Z0-9@$!%*#?&,.\-+()\[\]{}_\/]){8,32}$/; // eslint-disable-line
export const REGEX_EMAIL = /^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|((([a-zA-Z0-9]+[a-zA-Z\-0-9]+[a-zA-Z0-9]\.)|([a-zA-Z0-9]{1,}\.))+[a-zA-Z]{2,10}))$/; // eslint-disable-line
export const REGEX_NICKNAME = /^([a-zA-Z])[a-zA-Z0-9_-]{0,31}$/; // eslint-disable-line
export const REGEX_NAME = /^([a-zA-Z])[a-zA-Z0-9]{0,63}$/; // eslint-disable-line
