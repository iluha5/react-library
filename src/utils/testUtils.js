import checkPropTypes from 'check-prop-types';
import { applyMiddleware, createStore } from 'redux';
import rootReducer from 'reducers';
import { middlewares } from 'store';

export const findByTestAtrr = (component, attr) =>  component.find(`[data-test='${attr}']`);

export const findByClassName = (component, className) => component.find(`.${className}`);

export const checkProps = (component, expectedProps) =>
    checkPropTypes(component.propTypes, expectedProps, 'props', component.name);

export const testStore = (initialState) => {
    const createStoreWithMiddleware = applyMiddleware(...middlewares)(createStore);
    return createStoreWithMiddleware(rootReducer, initialState);
};
