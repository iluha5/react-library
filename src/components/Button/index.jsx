import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';

import style from './style.scss';

function Button ({btnType, className, size, children,  ...rest}) {
    return (
        <button
            className={cn(
                style.Btn,
                style[`Btn_size_${size}`],
                style[`Btn_type_${btnType}`],
                className,
            )}
            {...rest}
        >
            <span className={style['Btn-Label']}>{children}</span>
        </button>
    );
}

Button.defaultProps = {
    btnType: 'primary',
    size: 'm',
    disabled: false,
};

Button.propTypes = {
    className: PropTypes.any,
    btnType: PropTypes.oneOf(['primary', 'secondary', 'success', 'link', 'fake']),
    size: PropTypes.string,
    disabled: PropTypes.bool,
    children: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.object,
    ]),
    onClick: PropTypes.any,
};

export default Button;
