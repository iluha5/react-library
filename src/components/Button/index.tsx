import * as React from 'react';
import cn from 'classnames';

import style from './style.scss';

interface IProps extends React.HTMLProps<HTMLButtonElement> {
    btnType?: 'primary' | 'secondary' | 'success' | 'link' | 'fake',
    btnSize?: 'm' | 'xs',
    type?: 'button' | 'submit' | 'reset' | undefined
}

const Button: React.FC<IProps> = ({ btnType = 'primary', className, btnSize = 'm', children, ...rest }) => {
    return (
        <button
            className={cn(
                style.Btn,
                style[`Btn_size_${btnSize}`],
                style[`Btn_type_${btnType}`],
                className,
            )}
            {...rest}
        >
            <span className={style['Btn-Label']}>{children}</span>
        </button>
    );
};

export default Button;
