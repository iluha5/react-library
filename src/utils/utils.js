import { store } from 'store';
import { updateNotification } from 'ac/notifications';
import { REGEX_EMAIL } from 'utils/constants';

export const showNotification = (type, title) => {
    updateNotification(
        type,
        {
            title,
        }
    )(store.dispatch);
};


export const isEmailFn = (str) => {
    return REGEX_EMAIL.test(str.trim().toLowerCase());
};


