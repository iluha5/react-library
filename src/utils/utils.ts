import { store } from 'store/index';
import { updateNotification } from 'ac/notifications';
import { REGEX_EMAIL } from 'utils/constants';

export const showNotification = (type: string, title: string) => {
    updateNotification(
        type,
        {
            title,
        }
    )(store.dispatch);
};


export const isEmailFn = (str: string) => {
    return REGEX_EMAIL.test(str.trim().toLowerCase());
};

/**
 * Helper method, safely gets deep prop in the object
 * @param object - target object
 * @param path {string} - path to target prop in the object
 * @returns {*}
 * @example
 * const obj = {a: {b: 1}, c: null}
 * getDeepProp(obj, 'a.b')  // returns 1
 */
export function getDeepProp(object: object, path: string): any {
    const p = path.split('.');
    return p.reduce((xs: any, x: any) => (xs && xs[x] !== undefined ? xs[x] : undefined), object);
}




