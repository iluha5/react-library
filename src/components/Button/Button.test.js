import React from 'react';
import Button from 'components/Button';
import toJson from 'enzyme-to-json';

import {shallow} from 'enzyme';
import {checkProps, findByClassName} from 'utils/testUtils';

describe('Button component', () => {
    it('Should render without errors', () => {
        const component = shallow(<Button/>);
        expect(component.length).toBe(1);
    });

    it('Should not throw a warning', () => {
        const props = {
            className: 'TestingClass',
            btnType: 'secondary',
            size: 'm',
            disabled: false,
            children: 'Test button',
            onClick: () => {},
        };
        const propsError = checkProps(Button, props);
        expect(propsError).toBeUndefined();
    });

    describe('Renders', () => {
        let wrapper;
        let mockFunc;

        beforeEach(() => {
            mockFunc = jest.fn();
            const props = {
                children: 'Button text',
                onClick: mockFunc,
            };
            wrapper = shallow(<Button {...props} />);
        });

        it('Should emit callback on click event', () => {
            const button = findByClassName(wrapper, 'Btn');
            button.simulate('click');
            // console.log('wrapper', wrapper);

            const callback = mockFunc.mock.calls.length;
            expect(callback).toBe(1);
        });

        it('Render Button with default props', () => {
            const props = {
                children: 'Button text',
                onClick: mockFunc,
            };
            const but = toJson(shallow(<Button {...props} />));

            expect(but).toMatchSnapshot();
        });

        it('Render Button with size s, type secondary, ', () => {
            const props = {
                children: 'Button text',
                onClick: mockFunc,
                size: 's',
                btnType: 'secondary',
            };
            const but = toJson(shallow(<Button {...props} />));

            expect(but).toMatchSnapshot();
        });

        it('Render Button with size s, type success, ', () => {
            const props = {
                children: 'Button text',
                onClick: mockFunc,
                size: 's',
                btnType: 'success',
            };
            const but = toJson(shallow(<Button {...props} />));

            expect(but).toMatchSnapshot();
        });

        it('Render Button with size s, type fake, disabled ', () => {
            const props = {
                children: 'Button text',
                onClick: mockFunc,
                size: 's',
                btnType: 'fake',
                disabled: true,
            };
            const but = toJson(shallow(<Button {...props} />));

            expect(but).toMatchSnapshot();
        });

        it('Render Button with size m, type link', () => {
            const props = {
                children: 'Button text',
                onClick: mockFunc,
                size: 'm',
                btnType: 'link',
            };
            const but = toJson(shallow(<Button {...props} />));

            expect(but).toMatchSnapshot();
        });

        it('Render Button with size m, type link, classname mock-class', () => {
            const props = {
                className: 'mock-class',
                children: 'Button text',
                onClick: mockFunc,
                size: 'm',
                btnType: 'link',
            };
            const but = toJson(shallow(<Button {...props} />));

            expect(but).toMatchSnapshot();
        });
    });
});
