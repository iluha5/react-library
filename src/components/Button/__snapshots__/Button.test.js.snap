// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component Renders Render Button with default props 1`] = `
<button
  className="Btn Btn_size_m Btn_type_primary"
  disabled={false}
  onClick={[MockFunction]}
>
  <span
    className="Btn-Label"
  >
    Button text
  </span>
</button>
`;

exports[`Button component Renders Render Button with size m, type link 1`] = `
<button
  className="Btn Btn_size_m Btn_type_link"
  disabled={false}
  onClick={[MockFunction]}
>
  <span
    className="Btn-Label"
  >
    Button text
  </span>
</button>
`;

exports[`Button component Renders Render Button with size m, type link, classname mock-class 1`] = `
<button
  className="Btn Btn_size_m Btn_type_link mock-class"
  disabled={false}
  onClick={[MockFunction]}
>
  <span
    className="Btn-Label"
  >
    Button text
  </span>
</button>
`;

exports[`Button component Renders Render Button with size s, type fake, disabled  1`] = `
<button
  className="Btn Btn_size_s Btn_type_fake"
  disabled={true}
  onClick={[MockFunction]}
>
  <span
    className="Btn-Label"
  >
    Button text
  </span>
</button>
`;

exports[`Button component Renders Render Button with size s, type secondary,  1`] = `
<button
  className="Btn Btn_size_s Btn_type_secondary"
  disabled={false}
  onClick={[MockFunction]}
>
  <span
    className="Btn-Label"
  >
    Button text
  </span>
</button>
`;

exports[`Button component Renders Render Button with size s, type success,  1`] = `
<button
  className="Btn Btn_size_s Btn_type_success"
  disabled={false}
  onClick={[MockFunction]}
>
  <span
    className="Btn-Label"
  >
    Button text
  </span>
</button>
`;
